File: leetcode/p00705/p00705.go
package p00705

type MyHashSet struct {
	buckets [][]int
}

func hasher(key int) int {
	return ((key * 786433) & ((1 << 20) - 1)) >> 7
}

func contains(arr []int, key int) int {
	for i, num := range arr {
		if num == key {
			return i
		}
	}

	return -1
}

func Constructor() MyHashSet {
	return MyHashSet{
		buckets: make([][]int, 1<<13),
	}
}

func (this *MyHashSet) Add(key int) {
	hashed := hasher(key)
	if contains(this.buckets[hashed], key) == -1 {
		this.buckets[hashed] = append(this.buckets[hashed], key)
	}
}

func (this *MyHashSet) Remove(key int) {
	hashed := hasher(key)
	if idx := contains(this.buckets[hashed], key); idx > -1 {
		this.buckets[hashed] = append(this.buckets[hashed][:idx], this.buckets[hashed][idx+1:]...)
	}
}

func (this *MyHashSet) Contains(key int) bool {
	if contains(this.buckets[hasher(key)], key) == -1 {
		return false
	}

	return true
}

File: leetcode/p00705/p00705.py
class MyHashSet:
    def __init__(self):
        self.buckets = [[] for _ in range(1<<13)]

    def hasher(self, key):
        return ((key*786433) & (1<<20)-1) >> 7
    
    def add(self, key: int) -> None:
        hashed = self.hasher(key)
        if key not in self.buckets[hashed]:
            self.buckets[hashed].append(key)

    def remove(self, key: int) -> None:
        hashed = self.hasher(key)
        if key in self.buckets[hashed]:
            self.buckets[hashed].remove(key)

    def contains(self, key: int) -> bool:
        hashed = self.hasher(key)
        return key in self.buckets[hashed]

if __name__ == '__main__':
    obj = MyHashSet()
    
    obj.add(1)
    obj.add(2)
    assert obj.contains(1) == True
    assert obj.contains(3) == False
    obj.add(2)
    assert obj.contains(2) == True
    obj.remove(2)
    assert obj.contains(2) == False
